### ISSUE TITLE
Create an assertions or preconditions utility module

### ISSUE THREAD

**Issue by dwaynepryce** [microsoft/graspologic#558]
## Describe the solution you'd like
We do a lot of the same checks on our arguments; are you an int, are you a float, are you within a given range, etc.

It makes sense to write a function that will check this once and call it rather than write
```python
if not isinstance(foo, int):
  raise TypeError("Mr. T: Where's my int, foo")
if foo is not in range(low, high):
  raise ValueError("This is outside of the range we expect")
```

## Describe alternatives you've considered
Continue writing all of the same code we always have.

**Comment by dwaynepryce** [comment]
Also, if I've missed this in the code base feel free to toss a trout at me.  Part of this ticket was also to test the new issues templates :)

**Comment by bdpedigo** [comment]
i agree 💯 percent with this. 

Only other comment is sklearn has some array checking utils also that we can try to steal as much as possible too.

**Comment by dwaynepryce** [comment]
#571 will be able to make use of this as well

**Comment by bdpedigo** [comment]
some notes (in progress) on input checking that may be useful for us in sklearn. all of these are in the public API: 
- [check_array](https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_array.html#sklearn.utils.check_array)
- [check_scalar](https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_scalar.html#sklearn.utils.check_scalar)
   - I see absolutely no downside to using this one in particular in like a million places. allows for max and min value specs, prints errors based on a name for the parameter that you pass, etc.
- [check_consistent_length](https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_consistent_length.html#sklearn.utils.check_consistent_length)
- [column_or_1d](https://scikit-learn.org/stable/modules/generated/sklearn.utils.validation.column_or_1d.html#sklearn.utils.validation.column_or_1d)
- [check_random_state](https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_random_state.html#sklearn.utils.check_random_state)
- [check_symmetric](https://scikit-learn.org/stable/modules/generated/sklearn.utils.validation.check_symmetric.html#sklearn.utils.validation.check_symmetric)

**Comment by bdpedigo** [comment]
Dwayne: we should look into libraries with decorators that do this based on `typing`

**Comment by aaitbr** [comment]
hey @bdpedigo I am new, i want to contribute to this project, can i work on this ? With your guiding if it's possible ?

**Comment by bdpedigo** [comment]
hi @aaitbr, welcome! could you give us some info regarding your experience with open source (e.g. github) and python? Then we can help you get started on this. 

Do you mind if I ask whether you found our project on one of the sites with "good-first-issues" or somewhere else?

**Comment by aaitbr** [comment]
@bdpedigo yes sure ! so in python i'm familiar with libraries numpy, pandas & scikit-learn .. i know a little Spark, and i'm motivated and open to any learn in python to improve myself. Yes I found this project by using "good-first-issues" key, I read the documentation i liked it then i installed it.

**Comment by bdpedigo** [comment]
sounds good @aaitbr - 

as far as this issue goes, the goal is just to reduce the amount of input checking code we have across the library by refactoring that code to a utils or assertions module. 

I'm linking a few places you can see tons of input checking code:
https://github.com/microsoft/graspologic/blob/e465fa39ceb65156e3dea1d256fb566eb14aa215/graspologic/embed/base.py#L221

https://github.com/microsoft/graspologic/blob/e465fa39ceb65156e3dea1d256fb566eb14aa215/graspologic/inference/latent_distribution_test.py#L218

https://github.com/microsoft/graspologic/blob/e465fa39ceb65156e3dea1d256fb566eb14aa215/graspologic/simulations/simulations.py#L51

If you take this on, I'd recommend starting by focusing on `ValueError` checking - if you look above in this issue, we have discussed trying to set up `TypeError` checking once we get `typing` set up as well (see #504, which we could also use help on) using something like . Therefore, hopefully the type errors wont need checking at all if we can find a way to handle them with some sort of library that lets us just use decorators. 

I also link to a few input checking functions in sklearn that we can use for this project. 

As far as making PRs go, I'd recommend making small PRs (i.e. even if they only address one or two modules), because they will get reviewed faster and you'll get feedback faster. 

let me know if anything is unclear! like I said, #504 would also be another place we could use help - im not sure whether people have made much progress on it but I havent seen any PRs merged there.

**Comment by bradycargle** [comment]
Looks like this is still open, can I take a stab at it?

Have some experience in Python but am new to open source. To answer your earlier question @bdpedigo, I found this on one of the sites with "good-first-issues"

**Comment by bdpedigo** [comment]
I believe this is handled for the most part in #504, but @daxpryce will know better
